[Obs AI Assistant] Adds docs for connecting to a local LLM with the Obs AI Assistant (#2536)
elastic/docs-content: solutions/observability/connect-to-own-local-llm.md

mapped_pages:
  - https://www.elastic.co/guide/en/observability/current/connect-to-local-llm.html
applies_to:
  stack: ga
  serverless: ga
products:
  - id: observability
---

# Connect to your own local LLM

This page provides instructions for setting up a connector to a large language model (LLM) of your choice using LM Studio. This allows you to use your chosen model within the {{obs-ai-assistant}}. You’ll first need to set up LM Studio, then download and deploy a model via LM studio and finally configure the connector in your Elastic deployment.

::::{note}
You do not have to set up a proxy if LM studio is configured on the same network as your Elastic deployment or locally on your machine. 

If your Elastic deployment is not on the same network, you would need to configure a reverse proxy using Nginx to authenticate with Elastic. Refer [Configure your reverse proxy](https://www.elastic.co/docs/solutions/security/ai/connect-to-own-local-llm#_configure_your_reverse_proxy) for more detailed instructions.
::::

This example uses a server hosted in GCP to configure LM Studio with the [Mistral-Nemo-Instruct-2407](https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407) model.

> ## Configure LM Studio and download a model [_configure_lm_studio_and_download_a_model]

Reviewer: Do you think we could do more with referencing of the official docs from LM Studio itself instead of some step-by-step guides (e.g. selecting a model, downloading a model)? If the step-by-step instructions are specific to us, that's all right, but just cautious that this may become outdated if something changes in the LM Studio and we're not pointing to the "living" documentation.

Author: Some of these are generic commands and I linked the LM studio CLI docs as well.
However, loading the model with at least 64k context length is specific to us.

I feel like keeping this here will help a user to do an end-to-end set up rather than jumping on to different documentation (specially if they are doing it for the first time)

I'll add a section on top, for users who already have these configured to jump to the connector creation section in Kibana as @sorenlouv suggested [here](https://github.com/elastic/docs-content/pull/2536#discussion_r2272346974).

What do you think?

LM Studio supports the OpenAI SDK, which makes it compatible with Elastic’s OpenAI connector, allowing you to connect to any model available in the LM Studio marketplace.

As the first step, install [LM Studio](https://lmstudio.ai/).

> You must launch the application using its GUI before being able to use the CLI. For example, Chrome RDP with an [X Window System](https://cloud.google.com/architecture/chrome-desktop-remote-on-compute-engine) can be used for this purpose. After you’ve opened the application for the first time using the GUI, you can start the server by using `sudo lms server start` in the [CLI](https://lmstudio.ai/docs/cli/server-start).

Reviewer: [Suggestion] Maybe we should link to the link that's not specific to Google Cloud for setting up Chrome Remote Desktop, as it's a general-purpose remote access tool? For example, this [link](https://remotedesktop.google.com/) is for the tool itself, which might be enough of a pointer for the customers (also assuming that not all users wanting self-managed LLMs will use cloud-hosted VMs to host them).

Author: > also assuming that not all users wanting self-managed LLMs will use cloud-hosted VMs to host them

Makes sense. This part, I followed the same instructions as the security team's docs because at the top it's mentioned `This example uses a server hosted in GCP to configure LM Studio`

The tool doc that you linked has various options that's not relevant for this purpose I feel. My idea was to have a mechanism a completely new user to this process can follow without having to figure out too much by themselves.

Author: Thanks for the feedback @SrdjanLL 
Appreciate it

Finally, configure the connector:

1. Log in to your Elastic deployment.
2. Find the **Connectors** page in the navigation menu or use the [global search field](/explore-analyze/find-and-organize/find-apps-and-objects.md). Then click **Create Connector**, and select **OpenAI**. The OpenAI connector is compatible for this use case because LM Studio uses the OpenAI SDK.
3. Name your connector to help keep track of the model version you are using.
4. Under **Select an OpenAI provider**, select **Other (OpenAI Compatible Service)**.
5. Under **URL**, enter the host's IP address and port, followed by `/v1/chat/completions`. (If you have a reverse proxy set up, enter the domain name specified in your Nginx configuration file followed by `/v1/chat/completions`.)
6. Under **Default model**, enter `mistralai/mistral-nemo-instruct-2407`.
7. Under **API key**, fill in anything. (If you have a reverse proxy set up, enter the secret token specified in your Nginx configuration file.)
8. Click **Save**.

:::{image} /solutions/images/obs-ai-assistant-local-llm-connector-setup.png
:alt: The OpenAI create connector flyout

> :::

Reviewer 2: This is the most important part of the readme because this is specific to Elastic. Everything above is probably fine to include in the guide but we can assume that people will already know this, or can obtain the information elsewhere.

Author: Makes sense. I'll add a section on top for users who have already configured LM studio, to jump to the connector creation section.

Author: Added in https://github.com/elastic/docs-content/pull/2536/commits/8d53bb1a8ef510ac9f52efc1389e071b488753d2

---
navigation_title: Connect to a local LLM
mapped_pages:
  - https://www.elastic.co/guide/en/observability/current/connect-to-local-llm.html
applies_to:
  stack: ga
  serverless: ga
products:
  - id: observability
---

# Connect to your own local LLM

This page provides instructions for setting up a connector to a large language model (LLM) of your choice using LM Studio. This allows you to use your chosen model within the {{obs-ai-assistant}}. You’ll first need to set up LM Studio, then download and deploy a model via LM studio and finally configure the connector in your Elastic deployment.

::::{note}
You do not have to set up a proxy if LM studio is configured on the same network as your Elastic deployment or locally on your machine. 

> If your Elastic deployment is not on the same network, you would need to configure a reverse proxy using Nginx to authenticate with Elastic. Refer [Configure your reverse proxy](https://www.elastic.co/docs/solutions/security/ai/connect-to-own-local-llm#_configure_your_reverse_proxy) for more detailed instructions.

Author: For the reverse proxy, I referenced the docs by the security team as they have documented it. Let me know if I should document the same in this file instead of referring their document.

Reviewer: > you would need to configure a reverse proxy using Nginx to authenticate with Elastic

Is it a must to use Nginx for this purpose? If not, I'd suggest mentioning we have an example of setting up NGINX reverse proxy, but to be less prescriptive, because customers may be using other reverse proxies (HAProxy, Envoy Proxy, Caddy, etc.).

2. Find the **Connectors** page in the navigation menu or use the [global search field](/explore-analyze/find-and-organize/find-apps-and-objects.md). Then click **Create Connector**, and select **OpenAI**. The OpenAI connector is compatible for this use case because LM Studio uses the OpenAI SDK.
3. Name your connector to help keep track of the model version you are using.
4. Under **Select an OpenAI provider**, select **Other (OpenAI Compatible Service)**.
5. Under **URL**, enter the host's IP address and port, followed by `/v1/chat/completions`. (If you have a reverse proxy set up, enter the domain name specified in your Nginx configuration file followed by `/v1/chat/completions`.)
6. Under **Default model**, enter `mistralai/mistral-nemo-instruct-2407`.
7. Under **API key**, fill in anything. (If you have a reverse proxy set up, enter the secret token specified in your Nginx configuration file.)
8. Click **Save**.

:::{image} /solutions/images/obs-ai-assistant-local-llm-connector-setup.png
:alt: The OpenAI create connector flyout
:::

Setup is now complete. You can use the model you’ve loaded in LM Studio to power Elastic’s generative AI features.

::::{note}

> While local (open-weight) LLMs offer greater privacy and control, they generally do not match the raw performance and advanced reasoning capabilities of proprietary models by LLM providers mentioned in [here](/solutions/observability/observability-ai-assistant.md)

Author: @mdbirnstiehl How can I link a specific section on the obs ai assistant page here? I want to link `## Set up the AI Assistant [obs-ai-set-up]` section

Reviewer 3: You can add #obs-ai-set-up at the end of the link e.g. /solutions/observability/observability-ai-assistant.md#obs-ai-set-up

:::

After the model loads, you should see the message `Model loaded successfully` in the CLI.

:::{image} /solutions/images/observability-ai-assistant-model-loaded.png
:alt: The CLI message that appears after a model loads
:::

To verify which model is loaded, use the `lms ps` command.

:::{image} /solutions/images/observability-ai-assistant-lms-ps-command.png
:alt: The CLI message that appears after running lms ps
:::

If your model uses NVIDIA drivers, you can check the GPU performance with the `sudo nvidia-smi` command.

### Option 2: Load a model using the GUI [option-2-load-a-model-using-the-gui]

Once the model is downloaded, it will appear in the "My Models" window in LM Studio.

:::{image} /solutions/images/observability-ai-assistant-lm-studio-my-models.png
:alt: My Models window in LM Studio with downloaded models
:::

1. Navigate to the Developer window.
2. Click on the "Start server" toggle on the top left. Once the server is started, you'll see the address and port of the server. The port will be defaulted to 1234.
3. Click on "Select a model to load" and pick the model `Mistral Nemo Instruct 2407` from the dropdown menu.

> 4. Navigate to the "Load" on the right side of the LM Studio window, to adjust the context window to 64,000. Reload the model to apply the changes.

Reviewer 3: ```suggestion
4. Navigate to the **Load** on the right side of the LM Studio window, to adjust the context window to 64,000. Reload the model to apply the changes.
```

::::{note}

> To enable other devices in the same network access the server, turn on "Serve on Local Network" via Settings.

Reviewer 3: ```suggestion
To enable other devices on the same network to access the server, go to **Settings** and turn on **Serve on Local Network**.
```

::::

:::{image} /solutions/images/obs-ai-assistant-lm-studio-load-model-gui.png
:alt: Loading a model in LM studio developer tab
:::

## Configure the connector in your Elastic deployment [configure-the-connector-in-your-elastic-deployment]

Finally, configure the connector:

1. Log in to your Elastic deployment.

> 2. Find the **Connectors** page in the navigation menu or use the [global search field](/explore-analyze/find-and-organize/find-apps-and-objects.md). Then click **Create Connector**, and select **OpenAI**. The OpenAI connector is compatible for this use case because LM Studio uses the OpenAI SDK.

Reviewer 3: ```suggestion
2. Find the **Connectors** page in the navigation menu or use the [global search field](/explore-analyze/find-and-organize/find-apps-and-objects.md). Then click **Create Connector**, and select **OpenAI**. The OpenAI connector works for this use case because LM Studio uses the OpenAI SDK.
```
